### PR TITLE
[INFRA-6259] feat: implement reproducible zip

### PR DESCRIPTION
**Pull request by ajaxbits**
Alters the zip method to  reset all file permissions and timestamps of the files to be zipped, prior to zipping.

The file modification time respects `$SOURCE_DATE_EPOC` (according to the [reproducible builds spec](https://reproducible-builds.org/docs/timestamps/)), and defaults to 1/1/1980. The file permissions are also updated to `644`, to enforce standardization.

More information on reproducible builds for zip files can be found [here](https://reproducible-builds.org/docs/archives/).

### CODE REVIEW

File: package_python_function/reproducible_zip.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    zinfo = copy(zinfo)
+    zinfo.date_time = date_time()
+    if zinfo.is_dir():
+        zinfo.external_attr = (0o40000 | DIR_MODE) << 16
+        zinfo.external_attr |= 0x10  # MS-DOS directory flag
+    else:
+        zinfo.external_attr = FILE_MODE << 16
+    return zinfo
+
+class ReproducibleZipFile(ZipFile):
```
Comment (BrandonLWhite) [inline package_python_function/reproducible_zip.py:45]: I think this can be done without hijacking ZipFile.  Can you try seeing if [ZipFile.open()](https://docs.python.org/3/library/zipfile.html#zipfile.ZipFile.open) will allow you to achieve this?  It should allow you to specify a ZipInfo directly, and then you get a file-like object you can write to using something like `shutil.copyfileobj`
Here is an example I found: https://stackoverflow.com/questions/8068180/adding-a-file-like-object-to-a-zip-file-in-python
Except you will pass your ZipInfo object instead of the filename string shown in that example

**Review by BrandonLWhite — COMMENTED** [review]
See if you can simplify by using ZipFile.open() 

Also, the nested-zip case needs to be handled too.  It doesn't have to be in this PR if you prefer to get the basic case working first.

File: package_python_function/reproducible_zip.py (hunk 1 above)
Comment (ajaxbits) [reply package_python_function/reproducible_zip.py:45]: amazing. Yea I'll totally give this a try!

File: package_python_function/packager.py [hunk 2]
```diff
@@ -40,14 +46,32 @@ def package(self) -> None:
...
+            return (1980, 1, 1, 0, 0, 0)
+
+        with zipfile.ZipFile(target_path, "w", zipfile.ZIP_DEFLATED) as zip_file:
+
             def zip_dir(path: Path) -> None:
                 for item in path.iterdir():
                     if item.is_dir():
                         zip_dir(item)
                     else:
+                        zinfo = zipfile.ZipInfo.from_file(item)
```
Comment (ajaxbits) [inline package_python_function/packager.py:66]: @BrandonLWhite would love your feedback here.

I initially thought I would have to do: 
```python
 zinfo = zipfile.ZipInfo.from_file(item, item.relative_to(self.input_path))
```

since `from_file` accepts an `arcname` variable that is that path inside the archive. And the original code had a `relative_to` call in it.

However, I added some tests, and they pass without the `arcname` specified. 

So I think we're good here? Or is there an edge case I'm missing?

Comment (BrandonLWhite) [reply package_python_function/packager.py:66]: So long as the path that gets stored in the ZIP file is relative to the input_path, then it's all good.
Perhaps zip_file.open is doing a `relative_to` or similar internally since it has both paths available to it.

File: tests/test_package_python_function.py [hunk 3]
```diff
@@ -34,4 +37,19 @@ def test_package_python_function(tmp_path: Path) -> None:
...
+    verify_dir = tmp_path / "verify"
+    verify_dir.mkdir()
+    with zipfile.ZipFile(zip_file, "r") as zip:
+        zip.extractall(verify_dir)
+        for file_info in zip.infolist():
+            mode = (file_info.external_attr >> 16) & 0xFFFF
+            assert mode == EXPECTED_FILE_MODE
+            assert file_info.date_time == EXPECTED_FILE_DATE_TIME
+
+        assert verify_dir / "project_1" / "__init__.py"
```
Comment (BrandonLWhite) [inline tests/test_package_python_function.py:52]: These asserts don't appear to actually checking anything.  Perhaps you intended to call .exists() ?

File: package_python_function/packager.py (hunk 2 above)
Comment (ajaxbits) [reply package_python_function/packager.py:66]: That's kind of what I'm thinking as well.

File: tests/test_package_python_function.py (hunk 3 above)
Comment (ajaxbits) [reply tests/test_package_python_function.py:52]: ack!!! no wonder it always passed 🤦🏻 yea I'll fix that

File: package_python_function/packager.py (hunk 2 above)
Comment (ajaxbits) [reply package_python_function/packager.py]: Turns out that I just needed to fix my tests
